File: exercises/capstone/ex10_swept_footprint/ex10_swept_footprint.cpp
// Exercise 10 -- validating the swept footprint (Step 9).
//
// Read exercises/capstone/ex10_swept_footprint/README.md first.
#include "planning/airport/taxi_planner.hpp"

namespace planning::airport {

const char* toString(SweepViolation::Kind kind) {
  // TODO(you): one phrase per kind, for the refusal messages.
  (void)kind;
  return "?";
}

SweepResult validateSweep(const ZoneLayer& layer, const AircraftModel& aircraft, const Path& path,
                          ZoneClass startZone, const PermissionSet& permissions, bool offGraph) {
  // TODO(you): step along the path in kSweepStep metres and, at every pose,
  // apply the four checks of Step 9.
  //
  // 1. Gear tracks.  All three tyres on load-bearing pavement.  Shoulders do
  //    not count -- report them separately, because "you clipped the shoulder"
  //    and "you drove into the grass" are different bugs.
  //
  // 2. Wingtip envelope.  Clear of structures by the ICAO separation for this
  //    aircraft, with the larger margin on aprons.
  //
  // 3. Holding positions and runways, tested against the *swept outline*, not
  //    the centreline.  aircraft.footprint(pose) gives the outline;
  //    segmentIntersectsPolygon() and polygonsOverlap() do the tests.  A
  //    runway carries two identifiers for one strip, so naming either one in
  //    permissions.enterableRunways opens it.
  //
  // 4. When offGraph is true, the whole sweep must stay in startZone.  Stand
  //    to apron is the one permitted exception: the stand exists to be left.
  //
  // An empty path -- what a capture-window merge produces -- is a pass, not a
  // failure.  And one report per distinct problem is enough: a long stretch
  // off the pavement should not produce four hundred identical violations.
  (void)layer;
  (void)aircraft;
  (void)path;
  (void)startZone;
  (void)permissions;
  (void)offGraph;
  return SweepResult{};
}

}  // namespace planning::airport

File: src/capstone/problems/ex10-swept-footprint.ts
// Exercise 10 -- validating the swept footprint (Step 9).
// Brief: docs/capstone/ex10-swept-footprint.md

import {
  type AircraftModel,
  boundsOf,
  distanceOutsidePolygon,
  footprintOf,
  gearPoints,
  leftWingtip,
  type Path,
  pathAt,
  type PermissionSet,
  policyFor,
  polygonsOverlap,
  rightWingtip,
  segmentIntersectsPolygon,
  type ZoneClass,
  type ZoneLayer,
} from '../../airport/index.js';
import { SWEEP_STEP, type SweepResult } from '../types.js';
import { zoneAt } from './ex01-zone-layer.js';

export function validateSweep(
  layer: ZoneLayer,
  aircraft: AircraftModel,
  path: Path,
  startZone: ZoneClass,
  permissions: PermissionSet,
  offGraph: boolean,
): SweepResult {
  // TODO(you): up to here the planner has reasoned about a CURVE.  This is the
  // first and only place it reasons about the AEROPLANE.
  //
  // Sample the path every SWEEP_STEP metres and, at each pose, run four checks:
  //
  //   1  the three tyres are on load-bearing pavement.  A shoulder gets its own
  //      violation kind, because "rolled onto the shoulder" and "left the
  //      pavement entirely" are different phone calls.  When offGraph is true
  //      they must also still be in startZone -- stand to apron is the one
  //      permitted exception, because a stand exists to be left.
  //   2  neither wingtip is within the ICAO separation of a structure.  Use
  //      apronWingtipMargin on apron and stand, wingtipMargin elsewhere.
  //   3  the swept OUTLINE crosses no holding position the clearance has not
  //      named.  Not the centreline: the nose is 12.6 m ahead of the reference
  //      point, and it is the nose that crosses the line first.
  //   4  the swept outline enters no runway or protected area the clearance has
  //      not named.  A runway has two identifiers for one strip; naming either
  //      opens it.
  //
  // Finally, collapse the report.  The same violation repeats at every sample
  // of a long stretch, and four hundred identical violations are not four
  // hundred pieces of information.
  //
  // An empty path is a pass: a capture-window merge plans no motion at all, and
  // there is nothing to sweep.
  void layer;
  void aircraft;
  void path;
  void startZone;
  void permissions;
  void offGraph;
  void SWEEP_STEP;
  void zoneAt;
  void boundsOf;
  void distanceOutsidePolygon;
  void footprintOf;
  void gearPoints;
  void leftWingtip;
  void rightWingtip;
  void pathAt;
  void policyFor;
  void polygonsOverlap;
  void segmentIntersectsPolygon;
  return { ok: false, violations: [] };
}
